test/java/seedu/addressbook/commands/AddCommandTest.java

```diff
@@ -0,0 +1,159 @@
+            return;
+        }
+        String error = String.format(
+                "An add command was successfully constructed with invalid input: %s %s %s %s %s %s %s %s",
+                name, phone, isPhonePrivate, email, isEmailPrivate, address, isAddressPrivate, tags);
+        fail(error);
+    }
+    
+    @Test
+    public void addCommand_validData_correctlyConstructed() throws IllegalValueException {
+        Set<String> tags = new HashSet<>();
+        AddCommand cmd = new AddCommand(Name.EXAMPLE, Phone.EXAMPLE, true, Email.EXAMPLE, false,
+                Address.EXAMPLE, true, tags); // should never throw
+        ReadOnlyPerson p = cmd.getPerson();
+        
+        assertEquals(p.getName().fullName, Name.EXAMPLE);
+        assertEquals(p.getPhone().value, Phone.EXAMPLE);
+        assertTrue(p.getPhone().isPrivate());
+        assertEquals(p.getEmail().value, Email.EXAMPLE);
+        assertFalse(p.getEmail().isPrivate());
+        assertEquals(p.getAddress().value, Address.EXAMPLE);
+        assertTrue(p.getAddress().isPrivate());
+        boolean isTagListEmpty = !p.getTags().iterator().hasNext();
+        assertTrue(isTagListEmpty);
+    }
+    
+    @Test
+    public void addCommand_emptyAddressBook_addressBookContainsPerson() {
+        Person p = generateTestPerson();
+        AddCommand cmd = new AddCommand(p);
+        AddressBook book = new AddressBook();
+        cmd.setData(book, LAST_SHOWN_LIST);
+        CommandResult res = cmd.execute();
+        UniquePersonList people = book.getAllPersons();
+        
+        assertTrue(people.contains(p));
+        assertEquals(people.immutableListView().size(), 1);
+        assertFalse(res.getRelevantPersons().isPresent());
+        assertEquals(res.feedbackToUser, String.format(AddCommand.MESSAGE_SUCCESS, p));
+    }
+    
+    @Test
+    public void addCommand_addressBookAlreadyContainsPerson_addressBookUnmodified()
+            throws DuplicatePersonException {
+        Person p = generateTestPerson();
+        AddressBook book = new AddressBook();
+        book.addPerson(p); // should never throw
+        AddCommand cmd = new AddCommand(p);
+        cmd.setData(book, LAST_SHOWN_LIST);
+        CommandResult res = cmd.execute();
+        
+        assertFalse(res.getRelevantPersons().isPresent());
+        assertEquals(res.feedbackToUser, AddCommand.MESSAGE_DUPLICATE_PERSON);
+        UniquePersonList people = book.getAllPersons();
+        assertTrue(people.contains(p));
+        assertEquals(people.immutableListView().size(), 1);
+    }
+    
+    private static Person generateTestPerson() {
+        try {
+            return new Person(new Name(Name.EXAMPLE), new Phone(Phone.EXAMPLE, false),
+                    new Email(Email.EXAMPLE, true), new Address(Address.EXAMPLE, false), new UniqueTagList());
+        } catch (IllegalValueException e) {
+            fail("test person data should be valid by definition");
+            return null;
+        }
+    }
+}
```

Comment:
@Neurrone Oh, one small thing -- can add a newline at the end of file? For consistency with the other source files in this repo.